src/algebra/square.lean

```diff
@@ -0,0 +1,201 @@
+
+-/
+
+variables {R : Type*}
+
+/-- An element of a monoid is squarefree if the only squares that
+  divide it are the squares of units. -/
+def square [monoid R] (r : R) : Prop := ∃ x : R, x * x = r
+
+@[simp]
+lemma square_one [comm_monoid R] : square (1 : R) := by { use 1, simp, }
+
+@[simp]
+lemma square_zero [monoid_with_zero R] : square (0 : R) :=
+by { use 0, simp, }
+
+@[simp]
+lemma irreducible.not_square [comm_monoid R] {x : R} (h : irreducible x) :
+  ¬square x :=
+begin
+  rintros ⟨y, hy⟩,
+  rcases h.is_unit_or_is_unit hy.symm with hu | hu,
+  { rw ← hy at h,
+    exact h.not_unit (hu.mul hu), },
+  { rw ← hy at h,
+    exact h.not_unit (hu.mul hu), },
+end
+
+@[simp]
+lemma prime.not_square [cancel_comm_monoid_with_zero R] {x : R} (h : prime x) :
+  ¬square x := h.irreducible.not_square
+
+/-- The product of squares is a square. For a partial converse, see `nat.square_mul` -/
+lemma square_mul_of_square_of_square {R : Type*} [comm_monoid R] {m n : R} :
+  square m → square n → square (m * n) :=
+begin
+  rintros ⟨c, hc⟩,
+  rintros ⟨d, hd⟩,
+  use (c * d),
+  have : c * d * (c * d) = (c * c) * (d * d),
+  { rw mul_assoc,
+    conv in (d * (c * d)) { rw [mul_comm, mul_assoc], },
+    rw ←mul_assoc, },
+  rw [this, hc, hd],
+end
+
+section factorization
+
+namespace nat
+
+instance : decidable_pred (square : ℕ → Prop)
+| n := decidable_of_iff' _ (nat.exists_mul_self n)
+
+lemma square_iff_factorization_even {m : ℕ} :
+square m ↔ ∀ (p : ℕ), even (m.factorization p) :=
+begin
+  by_cases hm_zero : m = 0,
+  { simp only [hm_zero, square_zero, factorization_zero, finsupp.coe_zero,
+      pi.zero_apply, forall_const, true_iff],
+    use 0,
+    exact mul_zero 2, },
+  { split,
+    { rintros ⟨c, hc⟩,
+      by_cases hc' : c = 0,
+      { rw [hc', mul_zero] at hc,
+        rw ←hc,
+        simp only [factorization_zero, finsupp.coe_zero, pi.zero_apply, forall_const],
+        use 0,
+        exact mul_zero 2, },
+      { intros p,
+        rw [←hc, nat.factorization_mul hc' hc', finsupp.coe_add, ←two_mul],
+        simp [even], }, },
+    { intros hp,
+      use m.factorization.prod (λ p i, p ^ (i / 2)),
+      rw ← finsupp.prod_mul,
+      conv { to_rhs, rw ← nat.factorization_prod_pow_eq_self hm_zero,},
+      apply finsupp.prod_congr,
+      intros x hx,
+      rcases hp x with ⟨c, hc⟩,
+      rw [←pow_add, ←two_mul, hc],
+      congr,
+      simp, }, },
+end
+
+/-- The property of being a square is multiplicative. The ← direction can be generalized
+to an arbitrary commutative monoid. See `square_mul_of_square_of_square`. Similarly,
+this lemma could be generalized to an arbitrary unique factorization domain, but we make use
+of `nat.factorization` in this proof. -/
+lemma square_mul {m n : ℕ} (hmn : m.coprime n) :
+  square (m * n) ↔ square m ∧ square n :=
+begin
+  split,
+  { rw [square_iff_factorization_even, square_iff_factorization_even],
+    rw square_iff_factorization_even,
+    intros hmn',
+    split,
+    { intros p,
```

Comment:
You should separate this proof into a separate lemma, e.g.
```lean
private lemma square_mul_aux {m n : ℕ} (hmn : m.coprime n) :
  square (m * n) → square m := sorry
```
and then use commutativity to quickly get the other.

Reply:
In fact, I can see use in this as a standalone lemma, and same goes for the right counterpart, so I don't think they should be `private`.

Reply:
Done. Hopefully I got the `.` notation correct :-)